bulldozer/signals.go

```diff
@@ -21,109 +21,322 @@ import (
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.CommentSubstrings, err = s.commentSubstringMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.PRBodySubstrings, err = s.prBodyMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.Branches, err = s.branchMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.BranchPatterns, err = s.branchPatternMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	logger := zerolog.Ctx(ctx)
+	logger.Debug().Msgf("matches count: %d", matches.size())
+	logger.Debug().Msgf("singals count: %d", s.size())
+
+	if matches.size() == s.size() {
+		return true, fmt.Sprintf("pull request matches all %s signals", tag), nil
+	}
+
+	return false, fmt.Sprintf("pull request does not match all %s signals", tag), nil
+}
+
+// MatchesAny returns true if the pull request meets one or more signals. It also
 // returns a description of the signal that was met. The tag argument appears
 // in this description and indicates the behavior (trigger, ignore) this
 // set of signals is associated with.
-func (s *Signals) Matches(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+func (s *Signals) MatchesAny(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	descriptions, matches, err := s.labelMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, _, err = s.commentMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, _, err = s.commentSubstringMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, _, err = s.prBodyMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, _, err = s.branchMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, _, err = s.branchPatternMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	return false, fmt.Sprintf("pull request does not match the %s", tag), nil
+}
+
+// labelMatches determines which label signals match the given PR. It returns:
+// - A list of descriptions for each matched signal
+// - A list of the matched signals
+func (s *Signals) labelMatches(ctx context.Context, pullCtx pull.Context, tag string) ([]string, Signal, error) {
 	logger := zerolog.Ctx(ctx)
+	matches := Signal{}
+	descriptions := []string{}
 
 	labels, err := pullCtx.Labels(ctx)
```

Comment:
Kind of unrelated, but I think we could also fast exit here if there aren't any signals present. Same goes for the other functions.